go/src/lib.rs

```diff
@@ -3,38 +3,72 @@
  */
 
 // TODO: Investigate using uniffi bindings for Go instead of cbindgen
-
 #![deny(unsafe_op_in_unsafe_fn)]
-
+use derivative::Derivative;
 use glide_core::client::Client as GlideClient;
 use glide_core::connection_request;
 use glide_core::errors;
 use glide_core::errors::RequestErrorType;
+use glide_core::request_type::RequestType;
 use glide_core::ConnectionRequest;
 use protobuf::Message;
+use redis::{RedisResult, Value};
+use std::slice::from_raw_parts;
 use std::{
     ffi::{c_void, CString},
-    os::raw::c_char,
+    mem,
+    os::raw::{c_char, c_double, c_long, c_ulong},
 };
 use tokio::runtime::Builder;
 use tokio::runtime::Runtime;
 
-/// Success callback that is called when a Redis command succeeds.
+/// The struct represents the response of the command.
+///
+/// It will have one of the value populated depending on the return type of the command.
+///
+/// The struct is freed by the external caller by using `free_command_response` to avoid memory leaks.
+#[repr(C)]
+#[derive(Derivative)]
+#[derivative(Debug, Default)]
+pub struct CommandResponse {
+    int_value: c_long,
+    float_value: c_double,
+    bool_value: bool,
+
+    // Below two values are related to each other.
+    // `string_value` represents the string.
+    // `string_value_len` represents the length of the string.
+    #[derivative(Default(value = "std::ptr::null_mut()"))]
+    string_value: *mut c_char,
+    string_value_len: c_long,
+
+    // Below three values are related to each other.
+    // `array_value` represents the array of strings.
+    // `array_elements_len` represents the length of each array element.
+    // `array_value_len` represents the length of the array.
+    #[derivative(Default(value = "std::ptr::null_mut()"))]
+    array_value: *mut *mut c_char,
+    #[derivative(Default(value = "std::ptr::null_mut()"))]
+    array_elements_len: *mut c_long,
+    array_value_len: c_long,
+}
+
+/// Success callback that is called when a command succeeds.
 ///
 /// The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. The callback should be offloaded to a separate thread in order not to exhaust the client's thread pool.
 ///
 /// `index_ptr` is a baton-pass back to the caller language to uniquely identify the promise.
-/// `message` is the value returned by the Redis command. The 'message' is managed by Rust and is freed when the callback returns control back to the caller.
-// TODO: Change message type when implementing command logic
+/// `message` is the value returned by the command. The 'message' is managed by Rust and is freed when the callback returns control back to the caller.
 // TODO: Consider using a single response callback instead of success and failure callbacks
```

Comment:
You can remove this TODO if you have a final decition to continue with 2 callbacks

Reply:
Removed.